pinot-core/src/main/java/com/linkedin/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java

```diff
@@ -285,8 +285,26 @@ protected boolean consumeLoop() {
 
       int indexedMessageCount = 0;
       int kafkaMessageCount = 0;
+      boolean canTakeMore = true;
       while (!_shouldStop && !endCriteriaReached() && msgIterator.hasNext()) {
-        // Get a batch of messages from Kafka
+        if (!canTakeMore) {
+          // The RealtimeSegmentImpl that we are pushing rows into has indicated that it cannot accept any more
+          // rows. This can happen in one of two conditions:
+          // 1. We are in INITIAL_CONSUMING state, and we somehow exceeded the max number of rows we are allowed to consume
+          //    for this row. Something is seriously wrong, because endCriteriaReached() should have returned true when
+          //    we hit the row limit.
+          //    Throw an exception.
+          //
+          // 2. We are in CATCHING_UP state, and we legally hit this error due to a Kafka failure scenarios where
```

Comment:
I would add a short note that this is only in the case of an unclean leader election.